// @xmcl/unzip/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// index.ts
var unzip_exports = {};
__export(unzip_exports, {
  filterEntries: () => filterEntries,
  getEntriesRecord: () => getEntriesRecord,
  open: () => open,
  openEntryReadStream: () => openEntryReadStream,
  readAllEntries: () => readAllEntries,
  readEntry: () => readEntry,
  walkEntries: () => walkEntries,
  walkEntriesGenerator: () => walkEntriesGenerator
});
module.exports = __toCommonJS(unzip_exports);
var import_yauzl = require("yauzl");
async function open(target, options = { lazyEntries: true, autoClose: false }) {
  return new Promise((resolve, reject) => {
    function handleZip(err, zipfile) {
      if (err || !zipfile) {
        reject(err ?? new Error("Cannot open zip!"));
      } else {
        resolve(zipfile);
      }
    }
    if (typeof target === "string") {
      (0, import_yauzl.open)(target, options, handleZip);
    } else if (target instanceof Buffer) {
      (0, import_yauzl.fromBuffer)(target, options, handleZip);
    } else {
      (0, import_yauzl.fromFd)(target, options, handleZip);
    }
  });
}
function openEntryReadStream(zip, entry, options) {
  return new Promise((resolve, reject) => {
    function handleStream(err, stream) {
      if (err || !stream) {
        reject(err);
      } else {
        resolve(stream);
      }
    }
    if (options) {
      zip.openReadStream(entry, options, handleStream);
    } else {
      zip.openReadStream(entry, handleStream);
    }
  });
}
async function readEntry(zip, entry, options) {
  const stream = await openEntryReadStream(zip, entry, options);
  const buffers = [];
  await new Promise((resolve, reject) => {
    stream.on("data", (chunk) => {
      buffers.push(chunk);
    });
    stream.on("end", resolve);
    stream.on("error", reject);
  });
  return Buffer.concat(buffers);
}
async function* walkEntriesGenerator(zip) {
  let ended = false;
  let error;
  let resume = () => {
  };
  let wait = new Promise((resolve) => {
    resume = resolve;
  });
  const entries = [];
  const onEntry = (e) => {
    entries.push(e);
    resume();
  };
  const onEnd = () => {
    ended = true;
    resume();
  };
  const onError = (e) => {
    error = e;
    resume();
  };
  zip.addListener("entry", onEntry).addListener("end", onEnd).addListener("error", onError);
  try {
    while (!ended) {
      if (zip.lazyEntries) {
        zip.readEntry();
      }
      await wait;
      if (error) {
        throw error;
      }
      while (entries.length > 0 && !ended) {
        ended = !!(yield entries.pop());
      }
      wait = new Promise((resolve) => {
        resume = resolve;
      });
    }
  } finally {
    zip.removeListener("entry", onEntry).removeListener("end", onEnd).removeListener("error", onError);
  }
}
async function filterEntries(zip, entries) {
  const bags = entries.map((e) => [e, void 0]);
  let remaining = entries.length;
  for await (const entry of walkEntriesGenerator(zip)) {
    for (const bag of bags) {
      if (typeof bag[0] === "string") {
        if (bag[0] === entry.fileName) {
          bag[1] = entry;
          remaining -= 1;
        }
      } else {
        if (bag[0](entry)) {
          bag[1] = entry;
          remaining -= 1;
        }
      }
      if (remaining === 0)
        break;
    }
  }
  return bags.map((b) => b[1]);
}
async function walkEntries(zip, entryHandler) {
  const itr = walkEntriesGenerator(zip);
  for await (const entry of itr) {
    const result = await entryHandler(entry);
    if (result) {
      break;
    }
  }
}
function getEntriesRecord(entries) {
  const record = {};
  for (const entry of entries) {
    record[entry.fileName] = entry;
  }
  return record;
}
async function readAllEntries(zipFile) {
  const entries = [];
  for await (const entry of walkEntriesGenerator(zipFile)) {
    entries.push(entry);
  }
  return entries;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  filterEntries,
  getEntriesRecord,
  open,
  openEntryReadStream,
  readAllEntries,
  readEntry,
  walkEntries,
  walkEntriesGenerator
});
