// @xmcl/system/dist/index.mjs
// index.ts
import { open, readEntry, readAllEntries } from "@xmcl/unzip";
import { access as saccess, stat as sstat, writeFile as swriteFile, readFile as sreadFile, readdir as sreaddir } from "fs";
import { promisify } from "util";
import { join, sep } from "path";

// system.ts
var FileSystem = class {
  /**
     * Get the url for a file entry. If the system does not support get url. This should return an empty string.
     */
  getUrl(name) {
    return "";
  }
  isClosed() {
    return false;
  }
  close() {
  }
  // extension methods
  async missingFile(name) {
    return this.existsFile(name).then((v) => !v);
  }
  async walkFiles(target, walker) {
    if (await this.isDirectory(target)) {
      const childs = await this.listFiles(target);
      for (const child of childs) {
        await this.walkFiles(this.join(target, child), walker);
      }
    } else {
      const result = walker(this.join(target));
      if (result instanceof Promise) {
        await result;
      }
    }
  }
};

// index.ts
var access = promisify(saccess);
var stat = promisify(sstat);
var writeFile = promisify(swriteFile);
var readFile = promisify(sreadFile);
var readdir = promisify(sreaddir);
async function openFileSystem(basePath) {
  if (typeof basePath === "string") {
    const fstat = await stat(basePath);
    if (fstat.isDirectory()) {
      return new NodeFileSystem(basePath);
    } else {
      const zip = await open(basePath);
      Object.assign(zip, {
        validateFileName: (e) => {
          const fileName = e.fileName;
          if (/^[a-zA-Z]:/.test(fileName) || /^\//.test(fileName)) {
            e.fileName = fileName.replace(/^[a-zA-Z]:/, "").replace(/^\//, "");
          }
          return null;
        }
      });
      const entries = await readAllEntries(zip).then(
        (es) => (
          // ignore entries with '..' in the path
          es.filter((e) => e.fileName.split("/").indexOf("..") === -1)
        )
      );
      const entriesRecord = {};
      for (const entry of entries) {
        entriesRecord[entry.fileName] = entry;
      }
      return new NodeZipFileSystem(basePath, zip, entriesRecord);
    }
  } else {
    const zip = await open(basePath);
    const entries = await readAllEntries(zip);
    const entriesRecord = {};
    for (const entry of entries) {
      entriesRecord[entry.fileName] = entry;
    }
    return new NodeZipFileSystem("", zip, entriesRecord);
  }
}
function resolveFileSystem(base) {
  if (typeof base === "string" || base instanceof Uint8Array || base instanceof Buffer) {
    return openFileSystem(base);
  } else {
    return Promise.resolve(base);
  }
}
var NodeFileSystem = class extends FileSystem {
  constructor(root) {
    super();
    this.root = root;
  }
  sep = sep;
  type = "path";
  writeable = true;
  join(...paths) {
    return join(...paths);
  }
  getUrl(name) {
    return `file://${this.join(this.root, name)}`;
  }
  isDirectory(name) {
    return stat(join(this.root, name)).then((s) => s.isDirectory());
  }
  writeFile(name, data) {
    return writeFile(join(this.root, name), data);
  }
  existsFile(name) {
    return access(join(this.root, name)).then(() => true, () => false);
  }
  readFile(name, encoding) {
    return readFile(join(this.root, name), { encoding });
  }
  listFiles(name) {
    return readdir(join(this.root, name));
  }
  cd(name) {
    this.root = join(this.root, name);
  }
};
var NodeZipFileSystem = class extends FileSystem {
  constructor(root, zip, entries) {
    super();
    this.zip = zip;
    this.entries = entries;
    this.fileRoot = root;
  }
  sep = "/";
  type = "zip";
  writeable = false;
  zipRoot = "";
  fileRoot;
  isClosed() {
    return !this.zip.isOpen;
  }
  close() {
    this.zip.close();
  }
  get root() {
    return this.fileRoot + (this.zipRoot.length === 0 ? "" : `/${this.zipRoot}`);
  }
  normalizePath(name) {
    if (name.startsWith("/")) {
      name = name.substring(1);
    }
    if (this.zipRoot !== "") {
      name = [this.zipRoot, name].join("/");
    }
    return name;
  }
  join(...paths) {
    return paths.join("/");
  }
  isDirectory(name) {
    name = this.normalizePath(name);
    if (name === "") {
      return Promise.resolve(true);
    }
    if (this.entries[name]) {
      return Promise.resolve(name.endsWith("/"));
    }
    if (this.entries[name + "/"]) {
      return Promise.resolve(true);
    }
    const entries = Object.keys(this.entries);
    return Promise.resolve(entries.some((e) => e.startsWith(name + "/")));
  }
  existsFile(name) {
    name = this.normalizePath(name);
    if (this.entries[name] || this.entries[name + "/"]) {
      return Promise.resolve(true);
    }
    const entries = Object.keys(this.entries);
    return Promise.resolve(entries.some((e) => e.startsWith(name + "/")));
  }
  async readFile(name, encoding) {
    name = this.normalizePath(name);
    const entry = this.entries[name];
    if (!entry) {
      throw new Error(`Not found file named ${name}`);
    }
    const buffer = await readEntry(this.zip, entry);
    if (encoding === "utf-8") {
      return buffer.toString("utf-8");
    }
    if (encoding === "base64") {
      return buffer.toString("base64");
    }
    return buffer;
  }
  listFiles(name) {
    name = this.normalizePath(name);
    return Promise.resolve([
      ...new Set(Object.keys(this.entries).filter((n) => n.startsWith(name)).map((n) => n.substring(name.length)).map((n) => n.startsWith("/") ? n.substring(1) : n).map((n) => n.split("/")[0]))
    ]);
  }
  cd(name) {
    if (name.startsWith("/")) {
      this.zipRoot = name.substring(1);
      return;
    }
    const paths = name.split("/");
    for (const path of paths) {
      if (path === ".") {
        continue;
      } else if (path === "..") {
        const sub = this.zipRoot.split("/");
        if (sub.length > 0) {
          sub.pop();
          this.zipRoot = sub.join("/");
        }
      } else {
        if (this.zipRoot === "") {
          this.zipRoot = path;
        } else {
          this.zipRoot += `/${path}`;
        }
      }
    }
  }
  async walkFiles(startingDir, walker) {
    startingDir = this.normalizePath(startingDir);
    const root = startingDir.startsWith("/") ? startingDir.substring(1) : startingDir;
    for (const child of Object.keys(this.entries).filter((e) => e.startsWith(root))) {
      if (child.endsWith("/")) {
        continue;
      }
      const result = walker(child);
      if (result instanceof Promise) {
        await result;
      }
    }
  }
};
export {
  FileSystem,
  openFileSystem,
  resolveFileSystem
};
